geometry/calipers_box.cpp

```diff
@@ -0,0 +1,143 @@
+#include "geometry/calipers_box.hpp"
+
+#include "geometry/bounding_box.hpp"
+#include "geometry/convex_hull.hpp"
+#include "geometry/line2d.hpp"
+#include "geometry/polygon.hpp"
+#include "geometry/segment2d.hpp"
+
+#include "base/assert.hpp"
+
+#include <algorithm>
+#include <array>
+#include <limits>
+
+using namespace std;
+
+namespace m2
+{
+namespace
+{
+static_assert(numeric_limits<double>::has_infinity, "");
+
+double const kInf = numeric_limits<double>::infinity();
+double const kEps = 1e-12;
+
+// Checks whether (p1 - p) x (p2 - p) >= 0.
+bool IsCCW(PointD const & p1, PointD const & p2, PointD const & p)
+{
+  return robust::OrientedS(p1, p2, p) > -kEps;
+}
+
+PointD Ort(PointD const & p) { return PointD(-p.y, p.x); }
+
+// For each facet of the |hull| calls |fn| with the smallest rectangle
+// containing the hull and with one side colliear to the facet.
+template <typename Fn>
+void ForEachRect(ConvexHull const & hull, Fn && fn)
+{
+  ASSERT_GREATER(hull.Size(), 2, ());
+
+  size_t j = 0, k = 0, l = 0;
+  for (size_t i = 0; i < hull.Size(); ++i)
+  {
+    auto const ab = hull.SegmentAt(i).Dir();
+
+    j = max(j, i + 1);
+    while (DotProduct(ab, hull.SegmentAt(j).Dir()) > kEps)
+      ++j;
+
+    k = max(k, j);
+    while (CrossProduct(ab, hull.SegmentAt(k).Dir()) > kEps)
+      ++k;
+
+    l = max(l, k);
+    while (DotProduct(ab, hull.SegmentAt(l).Dir()) < -kEps)
+      ++l;
+
+    auto const oab = Ort(ab);
+    array<Line2D, 4> const lines = {{Line2D(hull.PointAt(i), ab), Line2D(hull.PointAt(j), oab),
+                                     Line2D(hull.PointAt(k), ab), Line2D(hull.PointAt(l), oab)}};
+    vector<PointD> corners;
+    for (size_t u = 0; u < lines.size(); ++u)
+    {
+      for (size_t v = u + 1; v < lines.size(); ++v)
+      {
+        auto result = LineIntersector::Intersect(lines[u], lines[v], kEps);
+        if (result.m_type == LineIntersector::Result::Type::One)
+          corners.push_back(result.m_point);
+      }
+    }
+
+    ConvexHull rect(corners);
+    if (rect.Size() == 4)
+      fn(rect.Points());
+  }
+}
+}  // namespace
+
+CalipersBox::CalipersBox(vector<PointD> const & points) : m_points({})
+{
+  ConvexHull hull(points);
+
+  if (hull.Size() < 4)
+  {
+    m_points = hull.Points();
+    return;
+  }
+
+  double bestArea = kInf;
+  vector<PointD> bestPoints;
+  ForEachRect(hull, [&](vector<PointD> const & points) {
+    ASSERT_EQUAL(points.size(), 4, ());
+    double const area = GetPolygonArea(points.begin(), points.end());
+    if (area < bestArea)
+    {
+      bestArea = area;
+      bestPoints = points;
+    }
+  });
+
+  if (bestPoints.empty())
+  {
+    BoundingBox bbox(points);
```

Comment:
How can we get here?

Reply:
We're compiling our project with "fast-math" flag, therefore all assumptions about IEEE754 behavior may be ruined. So, this is just for safety that we leaving object in the valid state.

Reply:
Also, this may be the case when the precision of input data exceeds 1e-12.